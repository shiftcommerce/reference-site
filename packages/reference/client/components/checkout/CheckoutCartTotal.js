// Libraries
import { Component } from 'react'
import Sticky from 'react-stickyfill'
import Link from 'next/link'

// Lib
import { calculateCartSummary } from '../../lib/calculateCartSummary'
import { fixedPrice } from '../../lib/fixedPrice'

// Objects
import Button from '../../objects/Button'

class CheckoutCartTotal extends Component {
  validOrder (checkout, order) {
    const checkoutSteps = [ 'shippingAddress', 'shippingMethod', 'paymentMethod' ]
    const cardErrors = order.card_errors
    const allStepsCompleted = checkoutSteps.every((step) => (checkout[step].completed))

    if (!cardErrors && allStepsCompleted) {
      return allStepsCompleted
    } else { return false }
  }

  renderButtons () {
    const { checkout, onClick, order } = this.props
    const isValidOrder = this.validOrder(checkout, order)

    if (checkout.currentStep === 3) {
      return <>
        <div className='c-cart-summary__buttons'>
          <Link href='/'>
            <Button
              aria-label='Continue Shopping'
              label='Continue shopping'
              status='primary'
              size='lrg'
              className='c-cart-summary__buttons--continue'
              type='button'
            />
          </Link>
          <Button
            aria-label='Review Order'
            label='Review Your Order'
            size='lrg'
            className='c-cart-summary__buttons--proceed'
            type='button'
            status='primary'
            onClick={onClick}
          />
        </div >
      </>
    } else if (checkout.currentStep === 4) {
      return <>
        <div className='c-cart-summary-buttons'>
          <Link href='/'>
            <Button
              aria-label='Continue Shopping'
              label='Continue shopping'
              size='lrg'
              className='c-cart-summary__buttons--continue'
              status='primary'
              type='button'
            />
          </Link>
          <Button
            aria-label='Place Order'
            label='Place Order'
            size='lrg'
            className='c-cart-summary__buttons--proceed'
            type='button'
            status='primary'
            id='place_order'
            status={(isValidOrder ? 'primary' : 'disabled')}
            disabled={!isValidOrder}
            onClick={onClick}
          />
        </div >
      </>
    } else {
      return <>
        <div className='c-cart-summary-buttons'>
          <Link href='/'>
            <Button
              aria-label='Continue Shopping'
              label='Continue shopping'
              size='lrg'
              className='c-cart-summary__buttons--continue'
              status='primary'
              type='button'
            />
          </Link>
          <Button
            aria-label='Continue to Payment'
            label='Continue to Payment'
            size='lrg'
            className='c-cart-summary__buttons--proceed'
            type='button'
            status='primary'
            onClick={onClick}
          />
        </div >
      </>
    }
  }

  render () {
    const { cart, checkout, order } = this.props
    const totals = calculateCartSummary(cart, checkout)

    let shippingText = ''
    if (!checkout.shippingAddress.completed) {
      shippingText = 'Enter address'
    } else if (!checkout.shippingMethod.retail_price_inc_tax) {
      shippingText = 'Select shipping method'
    } else {
      shippingText = `Â£${fixedPrice(totals.shipping)}`
    }

    return (
      <>
      <Sticky>
        <div aria-label='Cart total summary' className='u-sticky  c-cart-summary'>
          <div className=''>
            <dl aria-label='Subtotal'>
              <dt> Total Price: </dt>
              <dd> &pound;{ fixedPrice(totals.subTotal) } </dd>
            </dl>
            <dl aria-label='Shipping cost'>
              <dt> Shipping costs: </dt>
              <dd> { shippingText } </dd>
            </dl>
            <dl aria-label='Total' className='u-bold'>
              <dt> TOTAL: </dt>
              <dd> &pound;{ fixedPrice(totals.total) } </dd>
            </dl>
            <dl>
              <dt className='c-cart-summary__VAT'>* Including VAT</dt>
            </dl>
            { order.paymentError &&
              <div className='c-checkout-cart-total__payment-error'>{ order.paymentError }</div>
            }
          </div>
        </div>
      </Sticky>
      { this.renderButtons() }
      </>
    )
  }
}

export default CheckoutCartTotal
