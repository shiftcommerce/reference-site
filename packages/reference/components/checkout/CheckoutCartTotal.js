// Libraries
import { Component } from 'react'
import { calculateCartSummary } from '../../lib/calculateCartSummary'

class CheckoutCartTotal extends Component {
  render () {
    const totals = calculateCartSummary(this.props.cart, this.props.checkout)

    // TODO: This can be replaced with 'shippingAddress.valid' once validations are in place
    let shippingText = ''
    if (!this.props.checkout.shippingAddress.collapsed) {
      shippingText = 'Enter address'
    } else if (!this.props.checkout.shippingMethod.cost) {
      shippingText = 'Select shipping method'
    } else {
      shippingText = `Â£${totals.shipping}`
    }

    return (
      <div aria-label='Cart total summary' className='o-checkout-cart-total'>
        <div className='o-checkout-cart-total__wrapper'>
          <dl aria-label='Subtotal'>
            <dt> Subtotal: </dt>
            <dd> &pound;{totals.subTotal} </dd>
          </dl>
          <dl aria-label='VAT'>
            <dt> VAT: </dt>
            <dd> &pound;{totals.tax} </dd>
          </dl>
          <dl aria-label='Shipping cost'>
            <dt> Shipping: </dt>
            <dd> {shippingText} </dd>
          </dl>
          <dl aria-label='Total' className='o-checkout-cart-total__total'>
            <dt> You Pay: </dt>
            <dd> <b>&pound;{totals.total}</b> </dd>
          </dl>
        </div>
      </div>
    )
  }
}

export default CheckoutCartTotal
