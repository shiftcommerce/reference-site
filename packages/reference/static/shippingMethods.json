{
  "shippingMethods": [
    {
      "id": "1",
      "name": "Super Saver",
      "retail_price_inc_tax": 3.45,
      "paid_price_inc_tax": 3.45,
      "tax_amount": 2.00,
      "tax_percentage": 20.00,
      "sku": "super_saver",
      "reference": "super_saver",
      "delivery_date": "Sunday 2nd December"
    },
    {
      "id": "2",
      "name": "UPS Next Day",
      "retail_price_inc_tax": 5.45,
      "paid_price_inc_tax": 5.45,
      "tax_amount": 2.00,
      "tax_percentage": 20.00,
      "sku": "ups_next_day",
      "reference": "ups_next_day",
      "delivery_date": "Sunday 2nd December"
    },
    {
      "id": "3",
      "name": "UPS Next Day Before Noon",
      "retail_price_inc_tax": 8.45,
      "paid_price_inc_tax": 8.45,
      "tax_amount": 2.00,
      "tax_percentage": 20.00,
      "sku": "ups_next_day_before_noon",
      "reference": "ups_next_day_before_noon",
      "delivery_date": "Sunday 2nd December"
    },
    {
      "id": "4",
      "name": "UPS Same Day",
      "retail_price_inc_tax": 8.45,
      "paid_price_inc_tax": 8.45,
      "tax_amount": 2.00,
      "tax_percentage": 20.00,
      "sku": "ups_same_day",
      "delivery_date": "Sunday 2nd December",
      "reference": "ups_same_day"
    }
  ]
}
