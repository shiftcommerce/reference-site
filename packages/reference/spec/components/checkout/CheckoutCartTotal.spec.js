// Components
import CheckoutCartTotal from '../../../components/checkout/CheckoutCartTotal'

// Fixtures
import { shippingMethods } from '../../../static/shippingMethods.json'

test('renders the correct checkout cart total with single line item', () => {
  // arrange
  const cart = {
    lineItems: [
      {
        title: 'Test Product',
        price: 10,
        discount: 0,
        quantity: 2,
        sku: '123',
        imageUrl: '',
        size: 'size - 8'
      }
    ]
  }

  const checkout = {
    shippingAddress: {}
  }

  const updateQuantity = () => {}

  // act
  const wrapper = mount(
    <CheckoutCartTotal cart={cart} checkout={checkout} title="Checkout Cart Summary" />
  )

  // assert
  expect(wrapper).toMatchSnapshot()
  expect(wrapper).toIncludeText('Subtotal:  £20')
  expect(wrapper).toIncludeText('VAT:  £0')
  expect(wrapper).toIncludeText('Shipping:  Enter address')
  expect(wrapper).toIncludeText('You Pay:  £20')
})

test('renders the correct checkout cart total with multiple line items', () => {
  // arrange
  const cart = {
    lineItems: [
      {
        title: 'Test Product',
        price: 10,
        discount: 0,
        quantity: 2
      },
      {
        title: 'Pretend Product',
        price: 5,
        discount: 0,
        quantity: 1
      }
    ]
  }

  const checkout = {
    shippingAddress: {},
    shippingMethod: {}
  }

  const updateQuantity = () => {}

  // act
  const wrapper = mount(
    <CheckoutCartTotal cart={cart} checkout={checkout} />
  )

  // assert
  expect(wrapper).toMatchSnapshot()
  expect(wrapper).toIncludeText('Subtotal:  £25')
  expect(wrapper).toIncludeText('VAT:  £0')
  expect(wrapper).toIncludeText('Shipping:  Enter address')
  expect(wrapper).toIncludeText('You Pay:  £25')
})

test('renders the correct checkout cart costs with shipping address completed', () => {
  // arrange
  const cart = {
    lineItems: [
      {
        title: 'Test Product',
        price: 10,
        discount: 0,
        quantity: 2
      },
      {
        title: 'Pretend Product',
        price: 5,
        discount: 0,
        quantity: 1
      }
    ]
  }

  const checkout = {
    shippingAddress: {
      collapsed: true
    },
    shippingMethod: {}
  }

  const updateQuantity = () => {}

  // act
  const wrapper = mount(
    <CheckoutCartTotal cart={cart} checkout={checkout} />
  )

  // assert
  expect(wrapper).toMatchSnapshot()
  expect(wrapper).toIncludeText('Subtotal:  £25')
  expect(wrapper).toIncludeText('VAT:  £0')
  expect(wrapper).toIncludeText('Shipping:  Select shipping method')
  expect(wrapper).toIncludeText('You Pay:  £25')
})

test('renders the correct checkout cart costs with shipping method selected', () => {
  // arrange
  const cart = {
    lineItems: [
      {
        title: 'Test Product',
        price: 10,
        discount: 0,
        quantity: 2
      },
      {
        title: 'Pretend Product',
        price: 5,
        discount: 0,
        quantity: 1
      }
    ]
  }

  const checkout = {
    shippingAddress: {
      collapsed: true
    },
    shippingMethod: shippingMethods[0]
  }

  const updateQuantity = () => {}

  // act
  const wrapper = mount(
    <CheckoutCartTotal cart={cart} checkout={checkout} />
  )

  // assert
  expect(wrapper).toMatchSnapshot()
  expect(wrapper).toIncludeText('Subtotal:  £25')
  expect(wrapper).toIncludeText('VAT:  £0')
  expect(wrapper).toIncludeText('Shipping:  £3.45')
  expect(wrapper).toIncludeText('You Pay:  £28.45')
})
