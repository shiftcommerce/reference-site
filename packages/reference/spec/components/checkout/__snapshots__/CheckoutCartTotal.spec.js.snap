// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the correct checkout cart costs with shipping address completed 1`] = `
<CheckoutCartTotal
  cart={
    Object {
      "lineItems": Array [
        Object {
          "discount": 0,
          "price": 10,
          "quantity": 2,
          "title": "Test Product",
        },
        Object {
          "discount": 0,
          "price": 5,
          "quantity": 1,
          "title": "Pretend Product",
        },
      ],
    }
  }
  checkout={
    Object {
      "shippingAddress": Object {
        "collapsed": true,
      },
      "shippingMethod": Object {},
    }
  }
>
  <div
    aria-label="Cart total summary"
    className="o-checkout-cart-total"
  >
    <div
      className="o-checkout-cart-total__wrapper"
    >
      <dl
        aria-label="Subtotal"
      >
        <dt>
           Subtotal: 
        </dt>
        <dd>
           £
          25
           
        </dd>
      </dl>
      <dl
        aria-label="VAT"
      >
        <dt>
           VAT: 
        </dt>
        <dd>
           £
          0
           
        </dd>
      </dl>
      <dl
        aria-label="Shipping cost"
      >
        <dt>
           Shipping: 
        </dt>
        <dd>
           
          Select shipping method
           
        </dd>
      </dl>
      <dl
        aria-label="Total"
        className="o-checkout-cart-total__total"
      >
        <dt>
           You Pay: 
        </dt>
        <dd>
           
          <b>
            £
            25
          </b>
           
        </dd>
      </dl>
    </div>
  </div>
</CheckoutCartTotal>
`;

exports[`renders the correct checkout cart costs with shipping method selected 1`] = `
<CheckoutCartTotal
  cart={
    Object {
      "lineItems": Array [
        Object {
          "discount": 0,
          "price": 10,
          "quantity": 2,
          "title": "Test Product",
        },
        Object {
          "discount": 0,
          "price": 5,
          "quantity": 1,
          "title": "Pretend Product",
        },
      ],
    }
  }
  checkout={
    Object {
      "shippingAddress": Object {
        "collapsed": true,
      },
      "shippingMethod": Object {
        "cost": 3.45,
        "delivery_date": "Monday 3rd December",
        "id": "1",
        "title": "Super Saver",
        "value": "super_saver",
      },
    }
  }
>
  <div
    aria-label="Cart total summary"
    className="o-checkout-cart-total"
  >
    <div
      className="o-checkout-cart-total__wrapper"
    >
      <dl
        aria-label="Subtotal"
      >
        <dt>
           Subtotal: 
        </dt>
        <dd>
           £
          25
           
        </dd>
      </dl>
      <dl
        aria-label="VAT"
      >
        <dt>
           VAT: 
        </dt>
        <dd>
           £
          0
           
        </dd>
      </dl>
      <dl
        aria-label="Shipping cost"
      >
        <dt>
           Shipping: 
        </dt>
        <dd>
           
          £3.45
           
        </dd>
      </dl>
      <dl
        aria-label="Total"
        className="o-checkout-cart-total__total"
      >
        <dt>
           You Pay: 
        </dt>
        <dd>
           
          <b>
            £
            28.45
          </b>
           
        </dd>
      </dl>
    </div>
  </div>
</CheckoutCartTotal>
`;

exports[`renders the correct checkout cart total with multiple line items 1`] = `
<CheckoutCartTotal
  cart={
    Object {
      "lineItems": Array [
        Object {
          "discount": 0,
          "price": 10,
          "quantity": 2,
          "title": "Test Product",
        },
        Object {
          "discount": 0,
          "price": 5,
          "quantity": 1,
          "title": "Pretend Product",
        },
      ],
    }
  }
  checkout={
    Object {
      "shippingAddress": Object {},
      "shippingMethod": Object {},
    }
  }
>
  <div
    aria-label="Cart total summary"
    className="o-checkout-cart-total"
  >
    <div
      className="o-checkout-cart-total__wrapper"
    >
      <dl
        aria-label="Subtotal"
      >
        <dt>
           Subtotal: 
        </dt>
        <dd>
           £
          25
           
        </dd>
      </dl>
      <dl
        aria-label="VAT"
      >
        <dt>
           VAT: 
        </dt>
        <dd>
           £
          0
           
        </dd>
      </dl>
      <dl
        aria-label="Shipping cost"
      >
        <dt>
           Shipping: 
        </dt>
        <dd>
           
          Enter address
           
        </dd>
      </dl>
      <dl
        aria-label="Total"
        className="o-checkout-cart-total__total"
      >
        <dt>
           You Pay: 
        </dt>
        <dd>
           
          <b>
            £
            25
          </b>
           
        </dd>
      </dl>
    </div>
  </div>
</CheckoutCartTotal>
`;

exports[`renders the correct checkout cart total with single line item 1`] = `
<CheckoutCartTotal
  cart={
    Object {
      "lineItems": Array [
        Object {
          "discount": 0,
          "imageUrl": "",
          "price": 10,
          "quantity": 2,
          "size": "size - 8",
          "sku": "123",
          "title": "Test Product",
        },
      ],
    }
  }
  checkout={
    Object {
      "shippingAddress": Object {},
    }
  }
  title="Checkout Cart Summary"
>
  <div
    aria-label="Cart total summary"
    className="o-checkout-cart-total"
  >
    <div
      className="o-checkout-cart-total__wrapper"
    >
      <dl
        aria-label="Subtotal"
      >
        <dt>
           Subtotal: 
        </dt>
        <dd>
           £
          20
           
        </dd>
      </dl>
      <dl
        aria-label="VAT"
      >
        <dt>
           VAT: 
        </dt>
        <dd>
           £
          0
           
        </dd>
      </dl>
      <dl
        aria-label="Shipping cost"
      >
        <dt>
           Shipping: 
        </dt>
        <dd>
           
          Enter address
           
        </dd>
      </dl>
      <dl
        aria-label="Total"
        className="o-checkout-cart-total__total"
      >
        <dt>
           You Pay: 
        </dt>
        <dd>
           
          <b>
            £
            20
          </b>
           
        </dd>
      </dl>
    </div>
  </div>
</CheckoutCartTotal>
`;
